codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java

```diff
@@ -308,23 +315,91 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) t
+            }
+            return super.writeBytes(src);
+        }
+
+        Error writeError() {
+            return error;
+        }
+    }
+
     @Test
     public void releaseWhenMergeCumulateThrows() {
-        final Error error = new Error();
+        WriteFailingByteBuf oldCumulation = new WriteFailingByteBuf(1, 64);
+        ByteBuf in = Unpooled.buffer().writeZero(12);
+
+        Throwable thrown = null;
+        try {
+            ByteToMessageDecoder.MERGE_CUMULATOR.cumulate(UnpooledByteBufAllocator.DEFAULT, oldCumulation, in);
+        } catch (Throwable t) {
+            thrown = t;
+        }
+
+        assertSame(oldCumulation.writeError(), thrown);
+        assertEquals(0, in.refCnt());
+        assertEquals(1, oldCumulation.refCnt());
+    }
+
+    @Test
+    public void releaseWhenMergeCumulateThrowsInExpand() {
+        releaseWhenMergeCumulateThrowsInExpand(1, true);
+        releaseWhenMergeCumulateThrowsInExpand(2, true);
+        releaseWhenMergeCumulateThrowsInExpand(3, false); // sentinel test case
+    }
+
+    private void releaseWhenMergeCumulateThrowsInExpand(int untilFailure, boolean shouldFail) {
+        ByteBuf oldCumulation = UnpooledByteBufAllocator.DEFAULT.heapBuffer(8, 8);
+        final WriteFailingByteBuf newCumulation = new WriteFailingByteBuf(untilFailure, 16);
 
-        ByteBuf cumulation = new UnpooledHeapByteBuf(UnpooledByteBufAllocator.DEFAULT, 0, 64) {
+        ByteBufAllocator allocator = new AbstractByteBufAllocator(false) {
             @Override
-            public ByteBuf writeBytes(ByteBuf src) {
-                throw error;
+            public boolean isDirectBufferPooled() {
+                return false;
+            }
+
+            @Override
+            protected ByteBuf newHeapBuffer(int initialCapacity, int maxCapacity) {
+                return newCumulation;
+            }
+
+            @Override
+            protected ByteBuf newDirectBuffer(int initialCapacity, int maxCapacity) {
+                throw new UnsupportedOperationException();
             }
         };
+
         ByteBuf in = Unpooled.buffer().writeZero(12);
+        Throwable thrown = null;
         try {
-            ByteToMessageDecoder.MERGE_CUMULATOR.cumulate(UnpooledByteBufAllocator.DEFAULT, cumulation, in);
-            fail();
-        } catch (Error expected) {
-            assertSame(error, expected);
-            assertEquals(0, in.refCnt());
+            ByteToMessageDecoder.MERGE_CUMULATOR.cumulate(allocator, oldCumulation, in);
+        } catch (Throwable t) {
+            thrown = t;
+        }
+
+        assertEquals(0, in.refCnt());
+
+        if (shouldFail) {
+            assertSame(newCumulation.writeError(), thrown);
+            assertEquals(1, oldCumulation.refCnt());
+            assertEquals(0, newCumulation.refCnt());
+        } else {
+            assertNull(thrown);
+            assertEquals(0, oldCumulation.refCnt());
+            assertEquals(1, newCumulation.refCnt());
```

Comment:
you will need to call `newCumulation.release();` as well to ensure we not leak buffers in the test

Reply:
Oh ok! I assumed it wasn't necessary because it was from the unpooled heap and the test I started from didn't attempt to clean it up before. Will fix!